shell: fix unsafe API calls and add configurable autoflush behavior

jakub-uC opened the pull request:
Fixes an issue where the shell API could block indefinitely when called from threads other than the shell's processing thread, especially when the transport (e.g. USB CDC ACM) was unavailable or inactive.

Changes made:
1. Replaced `k_mutex_lock` calls with an indefinite timeout (`K_FOREVER`) by using a fixed timeout (`K_MSEC(SHELL_TX_MTX_TIMEOUT_MS)`) in shell API functions to prevent indefinite blocking.
2. Added a new Kconfig option `SHELL_PRINTF_AUTOFLUSH` to allow configurable autoflush behavior for shell printing functions.
3. Updated `Z_SHELL_FPRINTF_DEFINE` to use the `CONFIG_SHELL_PRINTF_AUTOFLUSH` setting instead of hardcoding the autoflush behavior to `true`.

Links: https://github.com/zephyrproject-rtos/zephyr/issues/84274

pdgendt commented on include/zephyr/shell/shell.h:
```diff
@@ -952,7 +952,7 @@ extern void z_shell_print_stream(const void *user_ctx, const char *data,
 	Z_SHELL_HISTORY_DEFINE(_name##_history, CONFIG_SHELL_HISTORY_BUFFER);                      \
 	Z_SHELL_FPRINTF_DEFINE(_name##_fprintf, &_name, _out_buf, CONFIG_SHELL_PRINTF_BUFF_SIZE,   \
-			       true, z_shell_print_stream);                                        \
+			       CONFIG_SHELL_PRINTF_AUTOFLUSH, z_shell_print_stream);               \
```
`CONFIG_SHELL_PRINTF_AUTOFLUSH` would expand to nothing if not enabled.

```suggestion
			       IS_ENABLED(CONFIG_SHELL_PRINTF_AUTOFLUSH), z_shell_print_stream);   \
```

jakub-uC replied:
Added, thank you

pdgendt reviewed (changes requested):
Can you split this into 2 commits?

1. addition of `CONFIG_SHELL_PRINTF_AUTOFLUSH`
2. Changing the timeout

jakub-uC committed:
shell: add Kconfig option for configurable autoflush behavior

Introduced a new Kconfig option `SHELL_PRINTF_AUTOFLUSH` to allow
configuring the autoflush behavior of shell printing functions.

Updated `Z_SHELL_FPRINTF_DEFINE` to use the
`CONFIG_SHELL_PRINTF_AUTOFLUSH` setting instead of hardcoding
the autoflush behavior to `true`.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@verkada.com>

jakub-uC committed:
shell: fix unsafe API calls and add configurable autoflush behavior

Fixes an issue where the shell API could block indefinitely when called
from threads other than the shell's processing thread, especially when
the transport (e.g. USB CDC ACM) was unavailable or inactive.

Replaced `k_mutex_lock` calls with an indefinite timeout (`K_FOREVER`)
by using a fixed timeout (`K_MSEC(SHELL_TX_MTX_TIMEOUT_MS)`) in shell
API functions to prevent indefinite blocking.

Link: https://github.com/zephyrproject-rtos/zephyr/issues/84274

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@verkada.com>

jakub-uC commented:
@pdgendt done

kartben reviewed (changes requested):
commit message should not include github references https://docs.zephyrproject.org/latest/contribute/guidelines.html#adding-links

henrikbrixandersen commented on subsys/shell/Kconfig:
```diff
@@ -94,6 +94,13 @@ config SHELL_PRINTF_BUFF_SIZE
+	bool "Indicate if the buffer should be automatically flushed"
+	default y
+	help
+	  Specify whether the shell's flash printing functions should automatically
```
What does "flash printing functions" cover here?

jakub-uC replied:
It's a mistake, I will remove it.

jakub-uC commented:
@kartben Fixed, I missed that new rule
